bootstrap/pkg/kfapp/gcp/gcp.go

```diff
@@ -672,10 +672,38 @@ func (gcp *Gcp) deleteEndpoints(ctx context.Context) error {
 	services := servicemanagement.NewServicesService(servicemanagementService)
 	op, err := services.Delete(gcp.Spec.Hostname).Context(ctx).Do()
 	if err != nil {
-		return &kfapis.KfError{
-			Code:    int(kfapis.INTERNAL_ERROR),
-			Message: fmt.Sprintf("issuing endpoint deletion error: %v", err),
+		nextPage := ""
+		// Use a loop to read multi-page managed services list.
+		for {
+			list := services.List().ProducerProjectId(gcp.Spec.Project)
+			if nextPage != "" {
+				list = list.PageToken(nextPage)
+			}
+			listResp, err := list.Do()
+			if err != nil {
+				return &kfapis.KfError{
+					Code:    int(kfapis.INTERNAL_ERROR),
+					Message: fmt.Sprintf("listing managed services error: %v", err),
+				}
+			}
+			for _, s := range listResp.Services {
+				if s.ServiceName == gcp.Spec.Hostname {
+					return &kfapis.KfError{
+						Code:    int(kfapis.INTERNAL_ERROR),
+						Message: fmt.Sprintf("issuing endpoint deletion error: %v", err),
+					}
+				}
+			}
+			// Explicitly copy it to prevent memory leak.
+			nextPage = "" + listResp.NextPageToken
+			if nextPage == "" {
+				break
+			}
 		}
+		// Delete is not successful and we are not able to find endpoint in managed
```

Comment:
Do you know why this happens?
delete failed, but the endpoint is gone

Reply:
probably ksonnet deleted it and the platform couldn’t. I think we’ll have various race conditions - eg deleting the namespace and deleting things in the namespace.

Reply:
yes, I think it's because ksonnet delete and gcp delete is trying to do overlapped tasks.